draft-ietf-tls-tls13.md

```diff
@@ -1953,11 +1947,14 @@ The "extension_data" field of this extension contains a
 %%% Key Exchange Messages
 
        struct {
-           select (role) {
-               case client:
+           select (Handshake.msg_type) {
+               case client_hello:
                    KeyShareEntry client_shares<0..2^16-1>;
 
-               case server:
+               case hello_retry_request:
+                   NamedGroup selected_group;
+
+               case server_hello:
                    KeyShareEntry server_share;
            }
        } KeyShare;
```

Comment:
Ugh, definitely not. supported_groups vs key_share is just syntax, but it must be a single group, not a list. The properties of the negotiation are complex enough as it is with the half-client half-server pick. We should at least ensure the server gets the final say.

For BoringSSL we've intentionally had the server just pick a group without paying attention to the key share list. Then we'll decide whether we HelloRetryRequest based on that decision. It's possible, in the future, there will be more variation in curve preferences but, right now, the only sensible curves are X25519 and P-256 anyway.

We already have extensions which are different between ServerHello and HelloRetryRequest. In fact, we have no extensions which are the same between the two. Every other extension is sent in at most one of the two, so implementations must already treat them different. Likewise, we already have extensions which are different between ClientHello and ServerHello. Also just fine.

I think the handful of cases where the same structure was used for client and server despite the two having very different roles (ALPN) were mistakes, and I'm glad we're not repeating them for 1.3's other extensions. No one is going to use the same code for client and server extension parsing because what you do with the fields is very different. It's purely an artificial constraint.